src/SIM/AdaptiveSIM.C

```diff
@@ -322,19 +322,29 @@ bool AdaptiveSIM::adaptMesh (int iStep)
   if (scheme == 2)
   {
     // Sum up the total error over all supported elements for each function
-    ASMbase* patch = model.getPatch(1);
-    if (!patch) return false;
     IntMat::const_iterator eit;
     IntVec::const_iterator nit;
-    for (i = 0; i < patch->getNoNodes(); i++) // Loop over basis functions
-      errors.push_back(DblIdx(0.0,i));
-    for (i = 1, eit = patch->begin_elm(); eit < patch->end_elm(); ++eit, i++)
-      for (nit = eit->begin(); nit < eit->end(); ++nit)
-        errors[*nit].first += eNorm(eRow,i);
+    int eOfs = 0;
+    errors.resize(model.getNoNodes(true));
+    for (ASMbase* patch : model.getFEModel()) {
+      for (i = 1, eit = patch->begin_elm(); eit < patch->end_elm(); eit++, i++)
+        for (nit = eit->begin(); nit < eit->end(); nit++) {
+          int node = patch->getNodeID(*nit+1)-1;
+          errors[node].first += eNorm(eRow,i+eOfs);
+          errors[node].second = node;
+        }
+      eOfs += patch->getNoElms();
+    }
   }
-  else
+  else {
+    if (model.getNoPatches() > 1)
+    {
+      std::cerr << "Multi-patch refinement only available for isotropic_function\n";
+      return false;
+    }
     for (i = 0; i < eNorm.cols(); i++)
       errors.push_back(DblIdx(eNorm(eRow,1+i),i));
+  }
```

Comment:
I would argue that this makes the logic less readable. The outermost if-else statement is "errors per function" ELSE "errors per element". The if-statement for single-patch is a temporary<sup>TM</sup> error-check since this particular refinement strategy is not yet implemented for multipatch models, but it might be at some later point in the future.

However since you brought this up, I've added comments to this part of the code to highlight this.